Signal/src/environment/Environment.m

```diff
@@ -171,69 +172,76 @@ - (PropertyListPreferences *)preferences
     return _preferences;
 }
 
-- (void)setHomeViewController:(HomeViewController *)signalsViewController
+- (void)setHomeViewController:(HomeViewController *)homeViewController
 {
-    _signalsViewController = signalsViewController;
+    _homeViewController = homeViewController;
 }
 
 - (void)setSignUpFlowNavigationController:(UINavigationController *)navigationController {
     _signUpFlowNavigationController = navigationController;
 }
 
-+ (void)messageThreadId:(NSString *)threadId {
-    TSThread *thread = [TSThread fetchObjectWithUniqueID:threadId];
-
-    if (!thread) {
-        DDLogWarn(@"We get UILocalNotifications with unknown threadId: %@", threadId);
-        return;
-    }
++ (void)presentConversationForRecipientId:(NSString *)recipientId
+{
+    [self presentConversationForRecipientId:recipientId keyboardOnViewAppearing:YES callOnViewAppearing:NO];
+}
 
-    if ([thread isGroupThread]) {
-        [self messageGroup:(TSGroupThread *)thread];
-    } else {
-        Environment *env          = [self getCurrent];
-        HomeViewController *vc = env.signalsViewController;
-        UIViewController *topvc   = vc.navigationController.topViewController;
-
-        if ([topvc isKindOfClass:[ConversationViewController class]]) {
-            ConversationViewController *mvc = (ConversationViewController *)topvc;
-            if ([mvc.thread.uniqueId isEqualToString:threadId]) {
-                [mvc popKeyBoard];
-                return;
-            }
-        }
-        [self messageIdentifier:((TSContactThread *)thread).contactIdentifier withCompose:YES];
-    }
++ (void)presentConversationForRecipientId:(NSString *)recipientId withCompose:(BOOL)compose
+{
+    [self presentConversationForRecipientId:recipientId keyboardOnViewAppearing:compose callOnViewAppearing:NO];
 }
 
-+ (void)messageIdentifier:(NSString *)identifier withCompose:(BOOL)compose {
-    Environment *env          = [self getCurrent];
-    HomeViewController *vc = env.signalsViewController;
++ (void)callRecipientId:(NSString *)recipientId
+{
+    [self presentConversationForRecipientId:recipientId keyboardOnViewAppearing:NO callOnViewAppearing:YES];
+}
 
-    [[TSStorageManager sharedManager].dbReadWriteConnection
-        asyncReadWriteWithBlock:^(YapDatabaseReadWriteTransaction *_Nonnull transaction) {
-            TSThread *thread = [TSContactThread getOrCreateThreadWithContactId:identifier transaction:transaction];
-            [vc presentThread:thread keyboardOnViewAppearing:YES callOnViewAppearing:NO];
-        }];
++ (void)presentConversationForRecipientId:(NSString *)recipientId
+                  keyboardOnViewAppearing:(BOOL)keyboardOnViewAppearing
+                      callOnViewAppearing:(BOOL)callOnViewAppearing
+{
+    DispatchMainThreadSafe(^{
```

Comment:
I used `DispatchMainThreadSafe()` in a couple of places in this class to ensure thread safety while avoiding dispatching async unnecessarily.